dttools/src/jx.c

```diff
@@ -0,0 +1,410 @@
+/*
+Copyright (C) 2015- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include "jx.h"
+#include "stringtools.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdarg.h>
+
+struct jx_pair * jx_pair( struct jx *key, struct jx *value, struct jx_pair *next )
+{
+	struct jx_pair *pair = malloc(sizeof(*pair));
+	pair->key = key;
+	pair->value = value;
+	pair->next = next;
+	return pair;
+}
+
+struct jx_item * jx_item( struct jx *value, struct jx_item *next )
+{
+	struct jx_item *item = malloc(sizeof(*item));
+	item->value = value;
+	item->next = next;
+	return item;
+}
+
+static struct jx * jx_create( jx_type_t type )
+{
+	struct jx *j = malloc(sizeof(*j));
+	j->type = type;
+	return j;
+}
+
+struct jx * jx_null()
+{
+	return jx_create(JX_NULL);
+}
+
+struct jx * jx_symbol( const char *symbol_name )
+{
+	struct jx *j = jx_create(JX_SYMBOL);
+	j->symbol_name = strdup(symbol_name);
+	return j;
+}
+
+struct jx * jx_string( const char *string_value )
+{
+	struct jx *j = jx_create(JX_STRING);
+	j->string_value = strdup(string_value);
+	return j;
+}
+
+struct jx * jx_string_nodup( char *string_value )
```

Comment:
The macro jx_format does exactly that.  (I would prefer jx_format to be a function but was running into some odd problems in passing a va_list down through the stack.  Maybe you can figure out how to make it a proper function.)

jx_string takes a plain string and we want to keep that option, otherwise someone will eventually get burned by trying to put a literal percent into a constant string.